File: LinkedList/Java/Problem9.java
/*
Reverse a DLL.
*/

/*package whatever //do not write package name here */

import java.io.*;

class DLL {
  int data;
  DLL next;
  DLL prev;

  DLL(int x) {
    data = x;
    next = null;
    prev = null;
  }
}

class GFG {
  public static DLL reverseDLL(DLL head) {
    if ((head == null) || (head.next == null))
      return head;
    DLL curr = head;
    DLL temp = null;
    while (curr != null) {
      temp = curr.prev;
      curr.prev = curr.next;
      curr.next = temp;
      curr = curr.prev;
    }
    return temp.prev;
  }

  public static DLL insertEnd(DLL head, int x) {
    DLL temp = new DLL(x);
    if (head == null) {
      return temp;
    }
    DLL curr = head;
    while (curr.next != null) {
      curr = curr.next;
    }
    curr.next = temp;
    temp.prev = curr;
    return head;
  }

  public static void printList(DLL node) {
    while (node != null) {
      System.out.print(node.data + " ");
      node = node.next;
    }
    System.out.println("");
  }

  public static void main(String[] args) {
    DLL head = null;
    for (int i = 10; i <= 100; i += 10)
      head = insertEnd(head, i);
    printList(head);
    head = reverseDLL(head);
    printList(head);
  }
}

File: LinkedList/Cpp/Problem9.cpp
/*
Reverse a DLL.
*/

#include <bits/stdc++.h>
using namespace std;

struct DLL
{
  int data;
  DLL *next;
  DLL *prev;
  DLL(int x)
  {
    data = x;
    next = NULL;
    prev = NULL;
  }
};

DLL *reverseDLL(DLL *head)
{
  if ((head == NULL) || (head->next == NULL))
    return head;
  DLL *curr = head;
  while (curr->next != NULL)
  {
    swap(curr->next, curr->prev);
    curr = curr->prev;
  }
  // curr exited at the last node.
  // So we swap for last node explicitly.
  swap(curr->next, curr->prev);
  return curr;
}

DLL *insertStart(DLL *head, int x)
{
  DLL *temp = new DLL(x);
  if (head == NULL)
  {
    return temp;
  }
  temp->next = head;
  head->prev = temp;
  return temp;
}

void printList(DLL *head)
{
  DLL *curr = head;
  while (curr != NULL)
  {
    cout << curr->data << " ";
    curr = curr->next;
  }
  cout << endl;
}

int main()
{
  DLL *head = NULL;
  for (int i = 10; i > 0; i--)
    head = insertStart(head, i);
  printList(head);
  head = reverseDLL(head);
  printList(head);
  return 0;
}
